scipy/stats/_continuous_distns.py

```diff
@@ -5206,6 +5206,321 @@ def _entropy(self, beta):
+        """
+        norm_values, shape_values = self._extract_positional_arguments(args)
+        return np.sum((norm * distribution.cdf(x, *shape) for norm, distribution, shape in zip(norm_values, self.distributions, shape_values)), axis=0)
+    
+    def _rvs(self, *args):
+        """
+        Generates random numbers using the individual distribution random generator
+        with a probability given by their individual norm factors 
+        """
+        norm_values, shape_values = self._extract_positional_arguments(args)
+        choices = np.random.choice(len(norm_values), size=self._size, p=norm_values)
+        result = np.zeros(self._size)
+        for i, (distribution, shape) in enumerate(zip(self.distributions, shape_values)):
+            mask = choices == i
+            result[mask] = distribution.rvs(size=mask.sum(), *shape)
+        return result
+
+    def _updated_ctor_param(self):
+        """
+        Set distributions as additional constructor argument
+        """
+        dct = super(mixture_gen, self)._updated_ctor_param()
+        dct['distributions'] = self.ctor_argument
+        return dct
+
+    def _argcheck(self, *args):
+        """
+        We allow for arbitrary arguments.
+        The sub distributions will check their arguments later anyway.
+        The default _argcheck method restricts the arguments to be positive.
+        """
+        return 1
+
+
+class template_gen(rv_continuous):
+    """
+    Generates a distribution given by a histogram.
+    This is useful to generate a template distribution from a binned datasample.
+
+    %(before_notes)s
+
+    Notes
+    -----
+    There are no additional shape parameters except for the loc and scale.
+    The pdf and cdf are defined as stepwise functions from the provided histogram.
+    In particular the cdf is not interpolated between bin boundaries and not differentiable.
+
+    %(after_notes)s
+
+    %(example)s
+
+    data = scipy.stats.norm.rvs(size=100000, loc=0, scale=1.5)
+    hist = np.histogram(data, bins=100)
+    template = scipy_extra.stats.template_gen(hist)
+
+    """
+    _support_mask = rv_continuous._support_mask
+
+    def __init__(self, histogram, *args, **kwargs):
+        """
+        Create a new distribution using the given histogram
+        @param histogram the return value of np.histogram
+        """
+        self.histogram = histogram
+        pdf, bins = self.histogram
+        bin_widths = (np.roll(bins, -1) - bins)[:-1]
+        pdf = pdf / float(np.sum(pdf * bin_widths)) 
+        cdf = np.cumsum(pdf * bin_widths)[:-1]
+        self.template_bins = bins
+        self.template_bin_widths = bin_widths
+        self.template_pdf = np.hstack([0.0, pdf, 0.0])
+        self.template_cdf = np.hstack([0.0, cdf, 1.0, 1.0])
+        # Set support
+        epsilon = 1e-7
+        kwargs['a'] = self.template_bins[0] - epsilon
+        kwargs['b'] = self.template_bins[-1] + epsilon
```

Comment:
This `epsilon` is confusing. I think it's better to define the distribution on a closed interval.